Use registry to find linked batteries for homekit

bdraco opened the pull request:
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Battery level is found in the following order:
1. Specified linked battery sensor via config (existing)
2. ATTR_BATTERY_LEVEL on the entity (existing)
3. Matched sensor with DEVICE_CLASS_BATTERY in the registry for the device

Battery charging state is found in the following order:
1. Specified linked battery charging sensor via config
2. ATTR_BATTERY_CHARGING on the Specified linked battery sensor via config (existing)
2. ATTR_BATTERY_CHARGING on the entity (existing)
3. Matched binary_sensor with DEVICE_CLASS_BATTERY_CHARGING in the registry for the device


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco committed:
Use registry to find linked batteries for homekit

balloob commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -61,6 +68,11 @@
     validate_media_player_features,
 )
 
+# TODO: switch the below once done testing outside of tree
+# from homeassistant.components.binary_sensor import DEVICE_CLASS_BATTERY_CHARGING
```
To do.

bdraco replied:
I'm going to do the binary sensor changes in a separate merge since it requires an architecture doc

balloob commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -357,12 +369,24 @@ def remove_bridge_accessory(self, aid):
+
+        bridged_states = []
+        for state in self.hass.states.async_all():
+            if not state or not self._filter(state.entity_id):
```
This can never be `not state`.

balloob commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -357,12 +369,24 @@ def remove_bridge_accessory(self, aid):
+        for state in self.hass.states.async_all():
+            if not state or not self._filter(state.entity_id):
+                continue
+            _async_configure_linked_battery_sensors(ent_reg, self._config, state)
```
Can we make this smarter? Now we're iterating over full device registry for each state. `O(stateCount * deviceRegistryCount)`

If we make a lookup table of devices and their device classes, we can run it in `O(stateCount + deviceRegistryCount)`.

bdraco replied:
As long as accessing `ent_reg.entities` directly is ok, I can definitely make this much more efficient.

balloob replied:
Yes that's ok. 

I am wondering here if we could make a reusable method to make a device lookup table with this structure:

```python
lookup[device_id][(domain, device_class)] = entity_id
```

bdraco replied:
@balloob I've done that. I'm not sure I've put it in the right place but its working as you suggest.

bdraco commented:
Will come back to this on the weekend when I have a larger chunk of time.

bdraco commented:
I'm going to come back to this once all the homekit PRs I have pending review are merged to avoid the conflict

bdraco commented:
Fixed all the conflicts, now going to do a break out of binary sensor for the arch change.  Then I'll add a lot more tests and optimize.

bdraco committed:
Fix conflict

bdraco committed:
freshen test

bdraco committed:
more cleanup

bdraco committed:
Increase battery test coverage

bdraco committed:
test to ensure we can find the linked battery sensors

bdraco committed:
Remove debug

bdraco committed:
tests for async_get_device_class_lookup

bdraco commented:
I wanted another device to test this with so I added vacuum support so I could test with my roomba https://github.com/home-assistant/core/pull/34386

bdraco committed:
reduce number of calls to async_add_executor_job, add constants

bdraco committed:
fix failing test

bdraco committed:
naming

bdraco committed:
Fix tests that broke due to merge conflict

bdraco committed:
Remove flags that were being used to
suppress homekit state updates after
changing state from homekit as
unexpected event was fixed
in HAP-python 2.7.0+

bdraco committed:
fix flakey homekit tests

bdraco commented:
/AzurePipelines run
